### ISSUE TITLE
RFC: Multimodal Request Flow Design (Stage 1)

### ISSUE THREAD

**Issue by zhengkezhou1** [sgl-project/sglang-jax#1390]
## Background

SGL-JAX previously routed multimodal requests and text-only requests through separate scheduling paths. This diverged from the unified architecture already adopted in the main SGLang repository, making it harder to share later batching, caching, and ViT-related optimizations. The goal of Stage 1 is to bring multimodal requests into the same scheduling path as text-only requests, align with the ViT processing direction in the main SGLang repository, and prepare the ground for a future JAX ViT forward path.

Qwen2.5-VL is the minimum verifiable model for this stage. The model uses ViT to split images into visual patches, inserts visual token placeholders into the text sequence, and then lets the LLM process the combined sequence. HuggingFace already provides mature preprocessing for this class of models, including dynamic image resizing, placeholder expansion, and M-RoPE position encoding. Stage 1 reuses that implementation directly to avoid introducing avoidable divergence.

### Stage 1 Scope

This stage covers only the first half of the request path: from the HTTP entrypoint to handing structured multimodal inputs to the Scheduler. Scheduler-side budget accounting, batch construction, and ViT/LLM forward computation are outside the scope of Stage 1 and will be addressed in later stages.

### Currently Unsupported Capabilities

To keep Stage 1 verifiable, the following capabilities are intentionally left for later stages:

- Audio and video modalities.
- Request-level parameters for dynamic image slicing, such as `max_dynamic_patch` and `min_dynamic_patch`.
- Multimodal passthrough for orthogonal advanced features, such as LoRA, disaggregated prefill, and custom logit processors.
- Upper bounds for visual patch counts produced by high-resolution images.

These missing capabilities do not affect ordinary single-image and multi-image chat requests.

## Data Flow

An image-bearing request passes through four stages. Each stage has a clear data-structure boundary so later stages can evolve independently.

```mermaid
flowchart LR
    A[HTTP request body<br/>messages + image_url] --> B[ChatCompletionRequest]
    B --> C[GenerateReqInput<br/>text + image_data]
    C --> D[TokenizedGenerateReqInput<br/>input_ids + MultimodalInputs]
    D --> E[(Scheduler)]

    subgraph S1[Stage 1 HTTP entrypoint]
        A --> B
    end
    subgraph S2[Stage 2 protocol-to-internal request]
        B --> C
    end
    subgraph S3[Stage 3 preprocessing and tokenize]
        C --> D
    end
    subgraph S4[Stage 4 handoff to Scheduler]
        D --> E
    end
```

### Stage 1: HTTP Entrypoint

The client sends a request to `/v1/chat/completions` using the OpenAI protocol. Images are carried as `image_url` content parts under `messages[*].content`. At the protocol layer, common inputs are remote URLs or base64 data URIs; after conversion to the internal request, the same loader can also support sources such as local paths.

- Input: HTTP request body.
- Output: a Pydantic-validated [ChatCompletionRequest](https://github.com/sgl-project/sglang-jax/blob/main/python/sgl_jax/srt/entrypoints/openai/protocol.py#L373), with multimodal content still preserved in the original content parts.

### Stage 2: Protocol Object to Internal Request

The [OpenAI Serving layer](https://github.com/sgl-project/sglang-jax/blob/main/python/sgl_jax/srt/entrypoints/openai/serving_chat.py#L50) converts the protocol object into the engine's internal generation request. This step splits content parts into a pure-text prompt and an ordered list of media references, then applies the model chat template to produce the final prompt with placeholders.

- Input: `ChatCompletionRequest`.
- Output: [GenerateReqInput](https://github.com/sgl-project/sglang-jax/blob/main/python/sgl_jax/srt/managers/io_struct.py#L276), containing the placeholder-bearing prompt text and per-modality media references.

### Stage 3: Preprocessing and Tokenize

[TokenizerManager](https://github.com/sgl-project/sglang-jax/blob/main/python/sgl_jax/srt/managers/tokenizer_manager.py#L117) is the unified preprocessing entrypoint. Text-only requests keep the existing tokenizer path unchanged. Requests with media are delegated to the modal processor, which performs media loading, HuggingFace preprocessing, placeholder expansion, and `input_ids` generation in one step.

For each image, the processor emits a separate multimodal item and computes an independent cache-disambiguation value for downstream radix cache keys. Model-specific position encodings, such as Qwen2.5-VL M-RoPE, are also computed at this stage so downstream components do not need to derive them again.

- Input: `GenerateReqInput`.
- Output: [TokenizedGenerateReqInput](https://github.com/sgl-project/sglang-jax/blob/main/python/sgl_jax/srt/managers/io_struct.py#L136), containing expanded `input_ids` and a typed multimodal payload, [`MultimodalInputs`](https://github.com/sgl-project/sglang-jax/blob/main/python/sgl_jax/srt/multimodal/common/modality_enum.py#L282).

### Stage 4: Handoff to Scheduler

`TokenizedGenerateReqInput` is sent to the [Scheduler](https://github.com/sgl-project/sglang-jax/blob/main/python/sgl_jax/srt/managers/scheduler.py#L130) through the existing inter-process channel. Stage 1 ends at this boundary: once the structured multimodal input is complete and available for downstream consumption, the Stage 1 objective is met. Scheduler-side logic is not covered in this document.

## Correctness Validation

Because both the reference and our request path rely on the same HuggingFace processor implementation, Stage 1 uses raw HuggingFace outputs as the baseline for end-to-end validation:

- The number of placeholders occupied by each image in the final `input_ids` matches the visual patch grid for that image.
- The number of visual feature rows for each image matches its patch grid; features for multiple images remain partitioned by image and do not overlap.
- Each multimodal item computes its cache-disambiguation value independently; different test images are distinguishable in radix cache keys.
- The M-RoPE position sequence after multi-image merging is exactly equal to the HuggingFace reference.
- The pixel tensor produced by running each image through the HuggingFace processor independently is byte-for-byte equal to the tensor stored at the corresponding position in our output, ruling out cross-image shuffling.

## Design Details

This section explains the abstractions introduced at the TokenizerManager and Processor layer in Stage 1. The goal is to decouple model-specific logic for turning media into model-consumable inputs from the main request path, while leaving extension points for later stages, including more models, more modalities, Scheduler work, and ViT forward. Readers do not need this section to understand the input/output contract of the data flow.

### Single Entrypoint and Routing

`TokenizerManager` is the unified entrypoint before requests enter the engine. It checks `contains_mm_input()` on the request object. If the request has no multimodal input, it follows the existing text-only tokenizer path and remains unaware of the multimodal changes. If the request contains media, it is delegated to the registered modal processor, which produces both expanded `input_ids` and structured multimodal metadata. Stage 1 intentionally avoids putting media splitting or model-specific field adaptation inside TokenizerManager, so the main path does not accumulate model-specialized logic.

### Processor Abstraction

`BaseMultimodalProcessor` defines a small contract:

- An abstract method, `process_mm_data_async(image_data, input_text, request_obj)`, must return an `(input_ids, MultimodalInputs)` pair. This is the only stable external API exposed by a processor.
- Model-agnostic helpers cover media-source normalization (URL, local path, `data:` URI, raw base64, `bytes`, PIL image, and numpy array), scalar-to-list normalization, and chat-template passthrough. These helpers live in the base class so each new model does not need to reimplement them.
- Remote fetching has a minimal safety boundary: a request timeout and a 64 MiB per-asset limit. These values are conservative constants in Stage 1; later stages can move them into `ServerArgs` and replace the implementation with asynchronous fetching.

```python
class BaseMultimodalProcessor(ABC):
    models: tuple[str, ...] = ()  # HF architectures supported by this processor

    def __init__(self, hf_config, server_args, processor):
        self.hf_config = hf_config
        self.server_args = server_args
        self.processor = processor  # underlying HF AutoProcessor

    def apply_chat_template(self, *args, **kwargs):
        return self.processor.apply_chat_template(*args, **kwargs)

    @abstractmethod
    async def process_mm_data_async(
        self, image_data, input_text, request_obj, **kwargs,
    ) -> tuple[list[int], "MultimodalInputs"]: ...

    @classmethod
    def load_image(cls, source) -> Image.Image: ...   # url / path / data uri / bytes

    @staticmethod
    def normalize_data(data) -> list: ...             # scalar -> [scalar], list passthrough
```

A model-specific implementation, such as `QwenVLProcessor`, only needs to declare the HuggingFace `architectures` it supports and implement `process_mm_data_async`. In Stage 1, that method directly reuses the same HuggingFace processor for preprocessing and then reshapes its output into the unified `MultimodalInputs` structure, preserving byte-for-byte parity with the reference.

### Automatic Registration by HF Architecture Name

`PROCESSOR_MAPPING` is a global table from "HF architecture name" to processor class. At startup, `import_processors` scans modules under the `processors/` package and registers every subclass of `BaseMultimodalProcessor` by each architecture name declared in its `models` tuple. During initialization, `get_mm_processor` looks up `hf_config.architectures` and instantiates the corresponding processor. Adding a new model only requires adding a new module and declaring `models = (...)`; TokenizerManager and the registry itself do not need to change.

```python
PROCESSOR_MAPPING: dict[str, type[BaseMultimodalProcessor]] = {}

def import_processors(package_name: str) -> None:
    # pkgutil-scan the processors package, register each subclass by its `models` tuple
    ...

def get_mm_processor(hf_config, server_args, processor) -> BaseMultimodalProcessor:
    for arch in hf_config.architectures:
        cls = PROCESSOR_MAPPING.get(arch)
        if cls is not None:
            return cls(hf_config, server_args, processor)
    raise ValueError(f"No multimodal processor registered for: {hf_config.architectures}")
```

### Unified Output Structure: `MultimodalInputs`

All processors converge on `MultimodalInputs`, so downstream components no longer need to know the concrete model. In this structure:

- `mm_items` is a list of `MultimodalDataItem` values split by media unit. Multiple images are represented as multiple independent items. Each item computes its own feature hash and `pad_value`, which downstream radix cache logic can use to distinguish cache keys and avoid false hits from treating multiple images as one merged item.
- Model-general fields, such as feature tensors, hashes, placeholder token IDs, and M-RoPE positions, live directly on dataclass fields. Model-specific fields, such as Qwen-VL `image_grid_thw`, are attached through the `model_specific_data` dictionary. This keeps the shared schema stable as new models are added.
- The structure also implements `merge`, reserving an interface for concatenating features, placeholders, and M-RoPE positions during future cross-request batching. Stage 1 does not depend on this path, but the semantic direction is already captured in the data structure.

```python
@dataclasses.dataclass
class MultimodalDataItem:
    modality: Modality
    hash: int | None = None
    pad_value: int | None = None
    feature: jax.Array | np.ndarray | None = None
    precomputed_embeddings: jax.Array | np.ndarray | None = None
    model_specific_data: dict[str, Any] = dataclasses.field(default_factory=dict)
    # e.g. QwenVLProcessor sets item["image_grid_thw"] = np.asarray([[t, h, w]])

@dataclasses.dataclass
class MultimodalInputs:
    mm_items: list[MultimodalDataItem]          # one item per image / video / audio
    im_token_id: int | None = None
    im_start_id: int | None = None
    im_end_id: int | None = None
    video_token_id: int | None = None
    audio_token_id: int | None = None
    mrope_positions: jax.Array | None = None    # Qwen-VL M-RoPE, precomputed here
    mrope_position_delta: jax.Array | None = None

    def merge(self, other): ...                 # reserved for cross-request batching
```

**Comment by JamesBrianD** [comment]
Whether the data in MultimodalDataItem MultimodalInputs is np or jax.numpy needs to be investigated. Excessive conversions between `jax.numpy` and `numpy` will significantly impact performance.

**Comment by JamesBrianD** [comment]
The RFC should spell out the concrete `TokenizerManager` migration.

Right now, when `server_args.multimodal` is enabled, sglang-jax does not initialize `model_config`, `context_len`, or `image_token_id`, and `mm_processor` remains `None`. The main `_tokenize_one_request()` path only runs regular text tokenization; it never calls `process_mm_data_async()` to produce processor-expanded `input_ids` and `mm_inputs`.

Please make the request flow explicit:

1. how `model_config/context_len/image_token_id` and `mm_processor` are initialized;
2. how `image_data/video_data/audio_data` are consumed during tokenization;
3. how context-length validation is done after the processor has produced the final `input_ids`;
4. how `mm_inputs` are attached to `TokenizedGenerateReqInput` and passed to the scheduler.

For reference, upstream SGLang handles this flow roughly as follows:

1. always initialize `model_config/context_len/image_token_id` during init;
2. if `model_config.is_multimodal`, register and create an `mm_processor`;
3. tokenize the initial text prompt first;
4. if the request contains image/audio/video data, call `mm_processor.process_mm_data_async(...)`;
5. if the processor returns `input_ids`, use them to replace the original `input_ids`;
6. validate context length against the final `input_ids`;
7. attach `mm_inputs` to `TokenizedGenerateReqInput` and send it to the scheduler;
8. let the scheduler handle padding, M-RoPE, and multimodal input materialization.

**Comment by JamesBrianD** [comment]
The cache-key semantics around multimodal items need a bit more clarification.

The RFC says each `MultimodalDataItem` has its own hash / `pad_value`, and that different images should be distinguishable by downstream cache logic. Since Stage 1 is focused on request processing rather than radix-cache implementation, please make the boundary explicit:

1. whether Stage 1 only produces stable per-image cache identity metadata, rather than implementing or validating radix-cache behavior;
2. what the item hash is computed from, e.g. raw image bytes, HF-processed pixel values, or another canonical representation;
3. whether `pad_value` is required to live outside the normal vocab-id range;
4. whether Stage 1 needs to carry per-image placeholder offsets, or whether that is deferred to the scheduler/cache stage.

If radix/prefix cache behavior is out of scope for Stage 1, the validation criteria should avoid claiming radix-cache correctness and instead validate only that each image produces independent, stable cache identity metadata.

**Comment by JamesBrianD** [comment]
Stage 2 owns message parsing and chat-template rendering. Its output is prompt text with symbolic media placeholders plus ordered media references. It does not own final token ids.

Stage 3 owns media preprocessing, placeholder expansion, and final input_ids generation. If the HF processor returns input_ids/padded_input_ids/M-RoPE metadata, TokenizerManager should preserve and forward those outputs instead of decode+retokenize.

**Comment by JamesBrianD** [comment]
Small wording concern: “visual feature rows” sounds like post-ViT embeddings, but Stage 1 is only covering preprocessing / scheduler handoff.

In SGLang, `MultimodalDataItem.feature` means processor output such as `pixel_values`, not the final visual embedding. The ViT runs later in model forward, and its output is then scattered into the text embeddings.

Could we phrase the validation in terms of processor outputs instead, e.g. `pixel_values`, `image_grid_thw`, per-image `MultimodalDataItem`, and placeholder counts? That would keep the Stage 1 boundary clear and avoid implying that ViT embedding validation is part of this RFC.
